ipld/amt/src/diff.rs

```diff
@@ -0,0 +1,398 @@
+            after: None,
+        });
+        Ok(true)
+    })?;
+
+    Ok(changes)
+}
+
+fn diff_leaves<Old, New>(
+    prev_node: &Node<Old>,
+    curr_node: &Node<New>,
+    offset: u64,
+) -> anyhow::Result<Vec<Change<Old, New>>>
+where
+    Old: Serialize + DeserializeOwned + Clone,
+    New: Serialize + DeserializeOwned + Clone,
+{
+    let prev_vals = match prev_node {
+        Node::Leaf { vals } => vals,
+        _ => {
+            anyhow::bail!("Leaf expected")
+        }
+    };
+
+    let curr_vals = match curr_node {
+        Node::Leaf { vals } => vals,
+        _ => {
+            anyhow::bail!("Leaf expected")
+        }
+    };
+
+    anyhow::ensure!(
+        prev_vals.len() == curr_vals.len(),
+        "node leaves have different numbers of values"
+    );
+
+    let mut changes = vec![];
+
+    for (i, (prev_val, curr_val)) in prev_vals.iter().zip(curr_vals.iter()).enumerate() {
+        let index = offset + i as u64;
+        match (prev_val, curr_val) {
+            (None, None) => continue,
+            (None, Some(curr_val)) => changes.push(Change {
+                key: index,
+                before: None,
+                after: Some(curr_val.clone()),
+            }),
+            (Some(prev_val), None) => changes.push(Change {
+                key: index,
+                before: Some(prev_val.clone()),
+                after: None,
+            }),
+            (Some(prev_val), Some(curr_val)) => {
+                if fvm_ipld_encoding::to_vec(&prev_val)? != fvm_ipld_encoding::to_vec(&curr_val)? {
+                    changes.push(Change {
+                        key: index,
+                        before: Some(prev_val.clone()),
+                        after: Some(curr_val.clone()),
+                    });
+                }
+            }
+        }
+    }
+
+    Ok(changes)
+}
+
+fn diff_node<Old, New, BS>(
+    prev_ctx: &NodeContext<BS>,
+    prev_node: &Node<Old>,
+    curr_ctx: &NodeContext<BS>,
+    curr_node: &Node<New>,
+    offset: u64,
+) -> anyhow::Result<Vec<Change<Old, New>>>
+where
+    Old: Serialize + DeserializeOwned + Clone,
+    New: Serialize + DeserializeOwned + Clone,
+    BS: Blockstore,
+{
+    if prev_ctx.height == 0 && curr_ctx.height == 0 {
+        diff_leaves(prev_node, curr_node, offset)
+    } else if curr_ctx.height > prev_ctx.height {
+        let mut changes = vec![];
+        let sub_count = curr_ctx.nodes_at_height();
+        let links = match curr_node {
+            Node::Link { links } => links,
+            _ => anyhow::bail!("Node::Link expected"),
+        };
+        for (i, link) in links.iter().enumerate() {
+            if let Some(link) = link {
+                let sub_ctx = NodeContext {
+                    height: curr_ctx.height - 1,
+                    bit_width: curr_ctx.bit_width,
+                    store: curr_ctx.store,
+                };
+                let sub_node = match link {
+                    node::Link::Cid { cid, .. } => sub_ctx
```

Comment:
Can we make this work with a dirty AMT?

Reply:
(and everywhere else)

Reply:
We'd either have to:

1. Flush.
2. Load and recurs.

The second one is probably the better option. Assume that the two nodes _may_ be different and continue.

Reply:
I've updated the logic in `diff` mod to support dirty nodes, however, in below code snippet `fvm_ipld_encoding::to_vec(node)` throws `Error: Serialization error for Cbor protocol: Msg("Tried to serialize without saving cache, run flush() on Amt before serializing")'` error, it seem that serializing dirty node(s) is unsupported. Any suggestions?
```rust
(Some(prev_link), Some(curr_link)) => {
                            let prev_cid = match prev_link {
                                node::Link::Cid { cid, .. } => *cid,
                                node::Link::Dirty(node) => Cid::new_v1(
                                    DAG_CBOR,
                                    multihash::Code::Blake2b256
                                        .digest(&fvm_ipld_encoding::to_vec(node)?),
                                ),
                            };
                            let curr_cid = match curr_link {
                                node::Link::Cid { cid, .. } => *cid,
                                node::Link::Dirty(node) => Cid::new_v1(
                                    DAG_CBOR,
                                    multihash::Code::Blake2b256
                                        .digest(&fvm_ipld_encoding::to_vec(node)?),
                                ),
                            };

                            if prev_cid == curr_cid {
                                continue;
                            }
```

Reply:
We can't serialize dirty intermediate nodes because we'd need the internal CIDs (which requires flushing). However, we _can_ simply assume that they may differ and recurs anyways.

Reply:
Basically:

1. If we have two CIDs, compare them. If they match, return.
2. In any other case, load the nodes (if not already dirty) and diff.

Reply:
@Stebalien  Done. Thx for the hints!